contrib/babelfishpg_tsql/src/catalog.c

```diff
@@ -3161,6 +3245,64 @@ grant_perms_to_objects_in_schema(const char *schema_name,
 
 			/* make sure later steps can see the object created here */
 			CommandCounterIncrement();
+}
+
+void
+grant_perms_to_objects_in_schema(const char *schema_name,
+				int16 priv,
+				const char *grantee)
+{
+	TableScanDesc scan;
+	Relation	bbf_schema_rel;
+	HeapTuple	tuple_bbf_schema;
+	const char	*object_name;
+	const char	*object_type;
+	ScanKeyData scanKey[3];
+	int16		dbid = get_cur_db_id();
+	const char *db_name = get_cur_db_name();
+	int16 permission = 0;
+
+	/* Fetch the relation */
+	bbf_schema_rel = table_open(get_bbf_schema_perms_oid(),
+									AccessShareLock);
+	ScanKeyInit(&scanKey[0],
+				Anum_bbf_schema_perms_dbid,
+				BTEqualStrategyNumber, F_INT2EQ,
+				Int16GetDatum(dbid));
+	ScanKeyInit(&scanKey[1],
+				Anum_bbf_schema_perms_schema_name,
+				BTEqualStrategyNumber, F_NAMEEQ,
+				CStringGetDatum(schema_name));
+	ScanKeyInit(&scanKey[2],
+				Anum_bbf_schema_perms_grantee,
+				BTEqualStrategyNumber, F_NAMEEQ,
+				CStringGetDatum(grantee));
+
+	scan = table_beginscan_catalog(bbf_schema_rel, 3, scanKey);
+	tuple_bbf_schema = heap_getnext(scan, ForwardScanDirection);
+
+	while (HeapTupleIsValid(tuple_bbf_schema))
+	{
+		Form_bbf_schema_perms schemaform;
+		schemaform = (Form_bbf_schema_perms) GETSTRUCT(tuple_bbf_schema);
+		object_name = pstrdup(NameStr(schemaform->object_name));
+		object_type = pstrdup(NameStr(schemaform->object_type));
+		permission = (priv & (schemaform->permission));
+		/* For each object, grant the permission explicitly. */
+		if (strcmp(object_name, "ALL") != 0)
+		{
+			if((permission & 32) == 32)
+				grant_perms_to_each_obj(db_name, object_type, schema_name, object_name, grantee, "execute");
+			if((permission & 16) == 16)
+				grant_perms_to_each_obj(db_name, object_type, schema_name, object_name, grantee, "select");
```

Comment:
16 means 10000 in boolean which means select privilege is to be granted.
so (permission & 16) == 16 means privileges list contains select privilege which needs to be granted.